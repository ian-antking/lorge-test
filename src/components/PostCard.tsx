import React from 'react'
import { Link } from 'react-router-dom'
import styled from 'styled-components'

import formatDate from '../utils/format-date'
import truncateText from '../utils/truncate-text'

interface Props {
  id: string
  title: string
  author: string
  blurb: string
  published: string
}

const StyledLink = styled(Link)`
  text-decoration: none;
  color: inherit;
`

const PostCardContainer = styled.div`
  width: 80vw;
  padding: 20px;
  border-radius: 5px;
  box-shadow: ${(props) => props.theme.shadow} 1.95px 1.95px 2.6px;
  margin: 20px;
  box-sizing: border-box;
`

const PostCardTitle = styled.div`
  font-weight: bold;
  font-size: 1.5em;
  margin-bottom: 10px;
`

const PostCardHeader = styled.div`
  display: flex;
  justify-content: flex-end;
  opacity: 0.5;
`

const PostCardContent = styled.div`
  margin: 10px;
  overflow: hidden;
  text-overflow: 'ellipsis';
`

const PostCardFooter = styled.div`
  display: flex;
  justify-content: flex-end;
  opacity: 0.5;
`

const PostCard: React.FC<Props> = ({
  id,
  title,
  author,
  blurb,
  published
}): React.ReactElement => (
  <StyledLink to={`/post/${id}`} key={id}>
    <PostCardContainer>
      <PostCardHeader>
        <span>{formatDate(published)}</span>
      </PostCardHeader>

      <PostCardTitle>
        <span>{title}</span>
      </PostCardTitle>

      <PostCardContent>{truncateText(blurb)}</PostCardContent>

      <PostCardFooter>
        <span>{author}</span>
      </PostCardFooter>
    </PostCardContainer>
  </StyledLink>
)

export default PostCard
