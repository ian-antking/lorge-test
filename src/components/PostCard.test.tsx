import { fireEvent, render } from '@testing-library/react'
import {
  MemoryRouter,
  unstable_HistoryRouter as HistoryRouter
} from 'react-router-dom'
import { createMemoryHistory } from 'history'
import PostCard from './PostCard'

import '@testing-library/jest-dom'

describe('PostCard', () => {
  it('renders the post title', () => {
    const { getByText } = render(
      <PostCard
        id="id"
        title="titleText"
        author="authorText"
        published={new Date(Date.now()).toISOString()}
        blurb="blurbText"
      />,
      { wrapper: MemoryRouter }
    )

    expect(getByText('titleText')).toBeTruthy()
  })

  it('renders the post author', () => {
    const { getByText } = render(
      <PostCard
        id="id"
        title="titleText"
        author="authorText"
        published={new Date(Date.now()).toISOString()}
        blurb="blurbText"
      />,
      { wrapper: MemoryRouter }
    )

    expect(getByText('authorText')).toBeInTheDocument()
  })

  it('renders the post blurb', () => {
    const { getByText } = render(
      <PostCard
        id="id"
        title="titleText"
        author="authorText"
        published={new Date(Date.now()).toISOString()}
        blurb="blurbText"
      />,
      { wrapper: MemoryRouter }
    )

    expect(getByText('blurbText')).toBeInTheDocument()
  })

  it('renders the post published date in the correct format', () => {
    const date = '2022-04-16T21:06:02.979Z'
    const { getByText } = render(
      <PostCard
        id="id"
        title="titleText"
        author="authorText"
        published={date}
        blurb="blurbText"
      />,
      { wrapper: MemoryRouter }
    )

    expect(getByText('16th April 2022')).toBeInTheDocument()
  })

  it('updates history when clicked', () => {
    const history = createMemoryHistory({
      initialEntries: ['/']
    })

    const { getByText } = render(
      <HistoryRouter history={history}>
        <PostCard
          id="id"
          title="titleText"
          author="authorText"
          published={new Date(Date.now()).toISOString()}
          blurb="blurbText"
        />
      </HistoryRouter>
    )

    fireEvent.click(getByText('titleText'))

    expect(history.location.pathname).toBe('/post/id')
  })

  it('renders consistently', () => {
    const { asFragment } = render(
      <PostCard
        id="id"
        title="titleText"
        author="authorText"
        published={new Date(Date.now()).toISOString()}
        blurb="blurbText"
      />,
      { wrapper: MemoryRouter }
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
