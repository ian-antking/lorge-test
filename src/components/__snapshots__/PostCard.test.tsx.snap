// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard renders consistently 1`] = `
<DocumentFragment>
  <a
    class="sc-bczRLJ eyjeUM"
    href="/post/id"
  >
    <div
      class="sc-gsnTZi cznRGl"
    >
      <div
        class="sc-hKMtZM hNxTqh"
      >
        <span>
          14th June 2022
        </span>
      </div>
      <div
        class="sc-dkzDqf jGIDpV"
      >
        <span>
          titleText
        </span>
      </div>
      <div
        class="sc-eCYdqJ eTHGuY"
      >
        blurbText
      </div>
      <div
        class="sc-jSMfEi jNHCZO"
      >
        <span>
          authorText
        </span>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
